apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/RetryingMutateRowsStream.cs

```diff
@@ -0,0 +1,128 @@
+﻿// Copyright 2018 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax.Grpc;
+using Grpc.Core;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+using static Google.Cloud.Bigtable.V2.BigtableMutateRowsRequestManager;
+
+namespace Google.Cloud.Bigtable.V2
+{
+    internal class RetryingMutateRowsStream : BigtableClient.MutateRowsStream
+    {
+        private readonly CallSettings _callSettings;
+        private readonly BigtableClient _client;
+        private readonly MutateRowsRequest _initialRequest;
+        private readonly BigtableClient.MutateRowsStream _initialStream;
+        private readonly RetryingResponseEnumerator _responseEnumerator;
+        private readonly RetrySettings _streamRetrySettings;
+
+        private BigtableClient.MutateRowsStream _currentStream;
+
+        internal RetryingMutateRowsStream(
+            BigtableClient client,
+            MutateRowsRequest initialRequest,
+            BigtableClient.MutateRowsStream initialStream,
+            CallSettings callSettings,
+            RetrySettings streamRetrySettings)
+        {
+            Debug.Assert(!(initialStream is RetryingMutateRowsStream));
+            Debug.Assert(callSettings.Timing?.Retry == null);
+
+            _client = client;
+            _initialRequest = initialRequest;
+            _initialStream = initialStream;
+            _currentStream = initialStream;
+            _callSettings = callSettings;
+            _streamRetrySettings = streamRetrySettings;
+
+            _responseEnumerator = new RetryingResponseEnumerator(this);
+        }
+
+        /// <inheritdoc/>
+        public override AsyncServerStreamingCall<MutateRowsResponse> GrpcCall => _currentStream.GrpcCall;
+
+        /// <inheritdoc/>
+        public override IAsyncEnumerator<MutateRowsResponse> ResponseStream => _responseEnumerator;
+
+        private class RetryingResponseEnumerator : IAsyncEnumerator<MutateRowsResponse>
```

Comment:
So even though this is shown as a stream, it's really just going to return a single value? I'm wondering whether we should look at a different approach to this, leaving the "raw" stream methods as they are, but adding an extra method or set of methods to do all of the retry parts and just returning a `Task<MutateRowsResponse>`. That would be easier for the client to consume, and would probably involve less worrying about getting "inside" the infrastructure.